designs/v1alpha4-api.md

```diff
@@ -0,0 +1,130 @@
+# v1alpha4 API Proposal
+This document proposes comprehensive Provisioner API improvements prior to the v0.4 release. Due to minor backwards incompatible changes, this will result in a API version bump (v1alpha3 -> v1alpha4). This is in accordance with [Kubernetes API Deprecation Policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/).
+
+These changes recommend:
+1. Cloud Provider specific extensions under `spec.provider`.
+2. Removal of `spec.cluster`.
+3. Pluralization of `spec.architecture` and `spec.operatingSystem`.
+4. Provisoning limits for maximum cpu, memory, etc, under `spec.limits`.
+
+## Strongly Typed Vendor Specific Fields
+Cloud Providers are currently limited to using well known `spec.labels` for configuration of vendor specific parameters. For example:
+- `spec.labels['node.k8s.aws/launch-template-name']`
+- `spec.labels['node.k8s.aws/subnet-name']`
+
+One benefit of this approach is that pods may use corresponding `spec.nodeSelector[...]` to request additional constraints on provisioned nodes. This information must be communicated through Kubernetes label values, which is awkward for the following use cases:
+1. Parameters are a list (e.g. `subnets: ["subneta", "subnetb", "subnetc"]`)
+2. Parameters are a struct (e.g. `tags: { "foo" : "bar" }`)
+3. Parameters do not comply with [the label value character set](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set). See: [#646](https://github.com/awslabs/karpenter/issues/646).
+
+We introduce a new `provider` field that enables strongly typed vendor specific parameters without violating vendor neutrality principles in the Karpenter codebase. We leverage Kubernetes `runtime.RawExtensions` to encapsulate these fields as raw bytes, which are then unmarshaled in vendor specific code. Vendors may implement arbitrary validation, defaulting, and provisioning behavior over the entire structure of these extensions. These structures are versioned separately from the Provisioner GVK to enable Cloud Providers to make backwards incompatible changes to provider specific configuration without requiring a version bump to the Provisioner CRD. If versioning is not specified by the user, it will be inferred and defaulted.
+
+```bash
+apiVersion: karpenter.sh/v1alpha4
+kind: Provisioner
+spec:
+  provider:
+    apiVersion: extensions.karpenter.sh/v1alpha1
+    kind: AWS
+    securityGroups: ["abc", "def"]
+    subnets: ["123", "456"]
+    launchTemplateName: "foo"
+```
+
+The Cloud Provider *may* continue to support corresponding well known labels at the pod level (e.g. `node.k8s.aws/subnet-name`).
```

Comment:
It is not clear whether this will hold true for the alternatives listed in later sections.

Reply:
In my mind, we'd be resetting most of these and just adding them at the provisioner level (for now).